### PR TITLE
core: use withOptions instead of withAffinity

### PR DESCRIPTION
**Pull request by lukaszx0**
Addresses #2738

cc @ejona86 @zhangkun83

### CODE REVIEW

**Comment by ejona86** [comment]
The build fails due to a deprecation warning.

File: core/src/main/java/io/grpc/CallOptions.java [hunk 1]
```diff
@@ -152,8 +152,10 @@ public Deadline getDeadline() {
 
   /**
    * Returns a new {@code CallOptions} with attributes for affinity-based routing.
+   *
+   * @deprecated use {@link #withOption(Key, Object)}.
    */
-  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/1766")
+  @Deprecated
```
Comment (ejona86) [inline core/src/main/java/io/grpc/CallOptions.java]: Go ahead and leave the `@ExperimentalApi` annotation, to show that we can actually delete it.

Comment (lukaszx0) [inline core/src/main/java/io/grpc/CallOptions.java]: Done.

**Comment by lukaszx0** [comment]
> The build fails due to a deprecation warning.

Fixed. PTAL.

**Comment by lukaszx0** [comment]
@ejona86 :shipit: ?

**Comment by lukaszx0** [comment]
No rush, but it'd be great to merge it and include in `1.2.x` release since "first half" of it has `PickSubchannelArgs` already landed.

**Comment by ejona86** [comment]
This doesn't seem exceptional enough to include in 1.2.x, especially since the 1.2.x binary creation would happen today-ish, other than #2801. Now that we're getting on a frequent release schedule, it doesn't seem to hurt too much to just delay.

**Comment by lukaszx0** [comment]
Ok, sounds good.

**Pull request merged**